File: camera/live_ws.go
package main

import (
	"context"
	"encoding/binary"
	"encoding/json"
	"log/slog"
	"net/http"
	"strings"
	"sync/atomic"
	"time"

	"nhooyr.io/websocket"
)

// liveWSMsg is a JSON control message on the camera↔server WebSocket.
type liveWSMsg struct {
	Type string `json:"type"`
}

// RunLiveRelay maintains a persistent WebSocket to the server and
// streams H.264 NAL units + Opus audio when a viewer is watching.
// It reconnects with exponential backoff on failure.
//
// The WebSocket stays open even when no viewer is watching (only
// keepalives flow). This gives instant WebRTC startup when a viewer
// connects. On battery-powered cellular cameras this prevents the
// modem from entering deep sleep (DRX/PSM) between uploads. A future
// battery-saver mode could instead establish the WebSocket on-demand
// via a start_stream command piggy-backed on the telemetry poll,
// trading ~10s of WebRTC startup latency for radio idle time.
func RunLiveRelay(ctx context.Context, client *Client, relay *LiveRelay) {
	backoff := 2 * time.Second
	const maxBackoff = 30 * time.Second

	for {
		err := runLiveWSSession(ctx, client, relay)
		if ctx.Err() != nil {
			return
		}
		if err != nil {
			slog.Debug("live relay WebSocket error", "err", err)
		}

		slog.Info("live relay reconnecting", "backoff", backoff)
		select {
		case <-ctx.Done():
			return
		case <-time.After(backoff):
		}
		backoff *= 2
		if backoff > maxBackoff {
			backoff = maxBackoff
		}
	}
}

func runLiveWSSession(ctx context.Context, client *Client, relay *LiveRelay) error {
	wsURL := buildWSURL(client.serverURL, client.deviceID)
	slog.Info("live relay connecting", "url", wsURL)

	headers := http.Header{}
	headers.Set("Authorization", "Bearer "+client.apiKey)

	conn, _, err := websocket.Dial(ctx, wsURL, &websocket.DialOptions{
		HTTPHeader: headers,
	})
	if err != nil {
		return err
	}
	defer conn.CloseNow()

	// Send ready message.
	readyMsg, _ := json.Marshal(liveWSMsg{Type: "ready"})
	if err := conn.Write(ctx, websocket.MessageText, readyMsg); err != nil {
		return err
	}

	slog.Info("live relay connected")

	// Read control messages from server in a goroutine.
	var streaming atomic.Bool
	controlCtx, controlCancel := context.WithCancel(ctx)
	defer controlCancel()

	go func() {
		defer controlCancel()
		for {
			_, data, err := conn.Read(controlCtx)
			if err != nil {
				return
			}
			var msg liveWSMsg
			if json.Unmarshal(data, &msg) != nil {
				continue
			}
			switch msg.Type {
			case "start_stream":
				slog.Info("live relay: viewer connected, starting stream")
				streaming.Store(true)
			case "stop_stream":
				slog.Info("live relay: no viewers, stopping stream")
				streaming.Store(false)
			}
		}
	}()

	// Main loop: forward frames when streaming is active.
	for {
		select {
		case <-controlCtx.Done():
			conn.Close(websocket.StatusNormalClosure, "closing")
			return controlCtx.Err()
		case frame, ok := <-relay.C():
			if !ok {
				conn.Close(websocket.StatusNormalClosure, "relay closed")
				return nil
			}
			if !streaming.Load() {
				continue // discard frames when no viewer is watching
			}
			if err := sendFrame(ctx, conn, frame); err != nil {
				return err
			}
		}
	}
}

// sendFrame writes a binary WebSocket frame with the wire format:
//
//	[4 bytes big-endian timestamp_ms] [1 byte flags] [payload]
//
// Flags:
//
//	bit 0: is_keyframe (video IDR)
//	bit 1: is_audio (1 = Opus, 0 = H.264)
func sendFrame(ctx context.Context, conn *websocket.Conn, frame LiveFrame) error {
	header := make([]byte, 5)
	binary.BigEndian.PutUint32(header[0:4], uint32(time.Now().UnixMilli()&0xFFFFFFFF))

	var flags byte
	if frame.IsKeyframe {
		flags |= 0x01
	}
	if frame.IsAudio {
		flags |= 0x02
	}
	header[4] = flags

	msg := make([]byte, 0, len(header)+len(frame.Data))
	msg = append(msg, header...)
	msg = append(msg, frame.Data...)

	return conn.Write(ctx, websocket.MessageBinary, msg)
}

// buildWSURL converts an HTTP server URL to a WebSocket URL for the
// live relay endpoint.
func buildWSURL(serverURL, deviceID string) string {
	base := strings.TrimRight(serverURL, "/")
	if strings.HasPrefix(base, "https://") {
		base = "wss://" + base[8:]
	} else if strings.HasPrefix(base, "http://") {
		base = "ws://" + base[7:]
	}
	return base + "/api/v1/cameras/" + deviceID + "/live"
}

File: ghostcam-py/ghostcam/live_ws.py
"""Live WebSocket relay to the server.

Mirrors camera/live_ws.go. Maintains a persistent ws:// or wss://
connection to /api/v1/cameras/{deviceID}/live, sends a `{"type":"ready"}`
JSON message after connect, then forwards LiveFrames as binary frames
when a viewer is watching.

Wire format (must match server/live.go):

  [4 bytes BE uint32 timestamp_ms (mod 2^32)]
  [1 byte flags: bit0 = is_keyframe, bit1 = is_audio]
  [payload — H.264 NAL unit OR Opus packet]

When no viewer is connected the server sends `{"type":"stop_stream"}`
and we drop frames silently. `{"type":"start_stream"}` flips streaming
back on. Reconnects with exponential backoff 2 → 30 s on disconnect.
"""

from __future__ import annotations

import asyncio
import json
import logging
import struct
import time

import websockets
from websockets.exceptions import ConnectionClosed

from ghostcam.client import Client
from ghostcam.live_relay import LiveRelay

logger = logging.getLogger(__name__)

INITIAL_BACKOFF = 2.0
MAX_BACKOFF = 30.0


def _build_ws_url(server_url: str, device_id: str) -> str:
    base = server_url.rstrip("/")
    if base.startswith("https://"):
        base = "wss://" + base[len("https://"):]
    elif base.startswith("http://"):
        base = "ws://" + base[len("http://"):]
    return f"{base}/api/v1/cameras/{device_id}/live"


async def run_live_relay(client: Client, relay: LiveRelay) -> None:
    """Persistent reconnect loop.

    Cancellation closes any active connection cleanly via the websockets
    library's context manager.
    """
    backoff = INITIAL_BACKOFF
    while True:
        try:
            await _run_one_session(client, relay)
            backoff = INITIAL_BACKOFF
        except asyncio.CancelledError:
            raise
        except Exception as e:  # noqa: BLE001
            logger.debug("live relay error: %s", e)
        logger.info("live relay reconnecting in %.0fs", backoff)
        await asyncio.sleep(backoff)
        backoff = min(backoff * 2, MAX_BACKOFF)


async def _run_one_session(client: Client, relay: LiveRelay) -> None:
    url = _build_ws_url(client.server_url, client.device_id)
    headers = {"Authorization": client._auth_header("GET", f"/api/v1/cameras/{client.device_id}/live")}
    logger.info("live relay connecting: %s", url)

    async with websockets.connect(
        url,
        additional_headers=headers,
        ping_interval=20,
        ping_timeout=20,
        max_size=None,  # binary frames may be large NAL units
    ) as conn:
        await conn.send(json.dumps({"type": "ready"}))
        logger.info("live relay connected")

        streaming = asyncio.Event()
        control_task = asyncio.create_task(
            _read_control_messages(conn, streaming),
            name="ws-control",
        )

        try:
            while True:
                frame = await relay.queue.get()
                if not streaming.is_set():
                    continue  # discard frames when no viewer
                payload = _pack_frame(frame.data, frame.is_keyframe, frame.is_audio)
                try:
                    await conn.send(payload)
                except ConnectionClosed:
                    return
        finally:
            control_task.cancel()
            try:
                await control_task
            except (asyncio.CancelledError, Exception):  # noqa: BLE001
                pass


async def _read_control_messages(
    conn,  # type: ignore[no-untyped-def]
    streaming: asyncio.Event,
) -> None:
    try:
        async for msg in conn:
            if isinstance(msg, bytes):
                continue
            try:
                obj = json.loads(msg)
            except json.JSONDecodeError:
                continue
            kind = obj.get("type")
            if kind == "start_stream":
                logger.info("live relay: viewer connected, starting stream")
                streaming.set()
            elif kind == "stop_stream":
                logger.info("live relay: no viewers, stopping stream")
                streaming.clear()
    except ConnectionClosed:
        return


def _pack_frame(data: bytes, is_keyframe: bool, is_audio: bool) -> bytes:
    """Wire format from camera/live_ws.go::sendFrame.

    [4 bytes BE uint32 ts_ms (mod 2^32)] [1 byte flags] [payload]
    """
    ts_ms = int(time.time() * 1000) & 0xFFFFFFFF
    flags = 0
    if is_keyframe:
        flags |= 0x01
    if is_audio:
        flags |= 0x02
    return struct.pack(">IB", ts_ms, flags) + data
